File: server/src/collectors/nasaFirmsJp.js
/**
 * NASA FIRMS — VIIRS/MODIS active-fire pixels over Japan.
 *
 * Auth: NASA_FIRMS_MAP_KEY (free; https://firms.modaps.eosdis.nasa.gov/api/area/).
 *
 * Endpoint:
 *   https://firms.modaps.eosdis.nasa.gov/api/area/csv/<MAP_KEY>/VIIRS_NOAA20_NRT/JPN/1
 *
 * Source: 'firms_active_fire' — geometry from per-row latitude/longitude.
 */

const TIMEOUT_MS = 20000;

const SENSOR = process.env.FIRMS_SENSOR || 'VIIRS_NOAA20_NRT';
const DAYS = Math.min(Number(process.env.FIRMS_DAYS || 1), 10);

export default async function collectNasaFirmsJp() {
  const key = process.env.NASA_FIRMS_MAP_KEY || process.env.FIRMS_MAP_KEY || '';
  if (!key) {
    return {
      type: 'FeatureCollection',
      features: [],
      _meta: {
        source: 'firms_no_key',
        fetchedAt: new Date().toISOString(),
        recordCount: 0,
        env_hint: 'Set NASA_FIRMS_MAP_KEY (free https://firms.modaps.eosdis.nasa.gov/api/area/)',
        description: 'NASA FIRMS active fires over JP — no key',
      },
    };
  }
  const url = `https://firms.modaps.eosdis.nasa.gov/api/area/csv/${key}/${SENSOR}/JPN/${DAYS}`;

  let csv = '';
  try {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), TIMEOUT_MS);
    const r = await fetch(url, { signal: ctrl.signal });
    clearTimeout(t);
    if (!r.ok) throw new Error(`HTTP ${r.status}`);
    csv = await r.text();
  } catch (err) {
    return {
      type: 'FeatureCollection',
      features: [],
      _meta: {
        source: 'firms_error',
        fetchedAt: new Date().toISOString(),
        recordCount: 0,
        error: err?.message || 'fetch_failed',
        description: 'NASA FIRMS — fetch failed',
      },
    };
  }

  const lines = csv.split(/\r?\n/);
  if (lines.length < 2) {
    return {
      type: 'FeatureCollection',
      features: [],
      _meta: {
        source: 'firms_empty',
        fetchedAt: new Date().toISOString(),
        recordCount: 0,
        description: 'NASA FIRMS — no fire pixels in window',
      },
    };
  }
  const header = lines[0].split(',');
  const idx = (name) => header.indexOf(name);
  const iLat = idx('latitude');
  const iLon = idx('longitude');
  const iBright = idx('bright_ti4') !== -1 ? idx('bright_ti4') : idx('brightness');
  const iAcq = idx('acq_date');
  const iAcqT = idx('acq_time');
  const iConf = idx('confidence');
  const iFrp = idx('frp');
  const iSat = idx('satellite');

  const features = [];
  for (let li = 1; li < lines.length; li += 1) {
    const line = lines[li].trim();
    if (!line) continue;
    const cols = line.split(',');
    const lon = Number(cols[iLon]); const lat = Number(cols[iLat]);
    const geocoded = Number.isFinite(lon) && Number.isFinite(lat);
    features.push({
      type: 'Feature',
      geometry: geocoded ? { type: 'Point', coordinates: [lon, lat] } : null,
      properties: {
        bright: Number(cols[iBright]),
        acq_date: cols[iAcq],
        acq_time: cols[iAcqT],
        confidence: cols[iConf],
        frp: Number(cols[iFrp]),
        satellite: cols[iSat],
        source: 'firms_active_fire',
      },
    });
  }

  return {
    type: 'FeatureCollection',
    features,
    _meta: {
      source: 'firms_active_fire',
      fetchedAt: new Date().toISOString(),
      recordCount: features.length,
      sensor: SENSOR,
      window_days: DAYS,
      env_hint: 'FIRMS_SENSOR (default VIIRS_NOAA20_NRT); FIRMS_DAYS (1-10)',
      description: 'NASA FIRMS — active-fire pixels over JP last N days',
    },
  };
}

File: native/collectors/sources/nasa_firms_jp.c
/* collectors/environment/sources/nasa_firms_jp.c — port of
 * server/src/collectors/nasaFirmsJp.js.
 * NASA FIRMS area CSV (requires NASA_FIRMS_MAP_KEY / FIRMS_MAP_KEY; free).
 * No key / fetch error / <2 lines → 0 features (SEED/_meta dropped, rule 7).
 * Header-indexed columns; Number() coercion → NaN becomes JSON null for the
 * numeric props (bright/frp); geometry null when lon/lat not finite.
 * Props order: bright, acq_date, acq_time, confidence, frp, satellite, source. */
#include "../../source.h"
#include "../../lib/feedlib.h"
#include "../../lib/geojson.h"
#include "../../lib/csv.h"
#include "../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define TIMEOUT_MS 20000

/* header.indexOf(name) — exact string match on a trimmed header cell. */
static int idx_of(cJSON *hdr, const char *name) {
  int i = 0;
  cJSON *c;
  cJSON_ArrayForEach(c, hdr) {
    if (cJSON_IsString(c) && strcmp(c->valuestring, name) == 0) return i;
    i++;
  }
  return -1;
}

static const char *cell(cJSON *row, int i) {
  if (i < 0) return NULL;
  cJSON *c = cJSON_GetArrayItem(row, i);
  return (c && cJSON_IsString(c)) ? c->valuestring : NULL;
}

/* JS Number(x): "" or non-numeric → NaN. We emit number or JSON null. */
static cJSON *num_or_null(const char *s) {
  if (!s || !*s) return cJSON_CreateNull();
  char *e;
  double v = strtod(s, &e);
  if (e == s) return cJSON_CreateNull();
  return cJSON_CreateNumber(v);
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  const char *key = getenv("NASA_FIRMS_MAP_KEY");
  if (!key || !*key) key = getenv("FIRMS_MAP_KEY");
  if (!key || !*key) {
    fprintf(stderr, "[nasa-firms-jp] no NASA_FIRMS_MAP_KEY — 0 rows\n");
    return 0;
  }
  const char *sensor = getenv("FIRMS_SENSOR");
  if (!sensor || !*sensor) sensor = "VIIRS_NOAA20_NRT";
  int days = 1;
  const char *ds = getenv("FIRMS_DAYS");
  if (ds && *ds) { int d = atoi(ds); days = d > 10 ? 10 : d; }

  char url[512];
  snprintf(url, sizeof url,
    "https://firms.modaps.eosdis.nasa.gov/api/area/csv/%s/%s/JPN/%d",
    key, sensor, days);

  char *csv = feed_get_text(ctx->http, url, TIMEOUT_MS);
  if (!csv) {
    fprintf(stderr, "[nasa-firms-jp] fetch failed\n");
    return -1;
  }

  /* csv.split(/\r?\n/) ; need >=2 lines (header + data). csv_parse with
   * headers=0 returns array of rows of string cells; row 0 is the header. */
  cJSON *rows = csv_parse(csv, 0);
  free(csv);
  int nrows = cJSON_GetArraySize(rows);
  if (nrows < 2) { cJSON_Delete(rows); return 0; }

  cJSON *hdr = cJSON_GetArrayItem(rows, 0);  /* exhaustive-ok: CSV header row, not a record */
  int iLat = idx_of(hdr, "latitude"), iLon = idx_of(hdr, "longitude");
  int iBright = idx_of(hdr, "bright_ti4");
  if (iBright == -1) iBright = idx_of(hdr, "brightness");
  int iAcq = idx_of(hdr, "acq_date"), iAcqT = idx_of(hdr, "acq_time");
  int iConf = idx_of(hdr, "confidence"), iFrp = idx_of(hdr, "frp");
  int iSat = idx_of(hdr, "satellite");

  cJSON *features = cJSON_CreateArray();
  for (int li = 1; li < nrows; li++) {
    cJSON *row = cJSON_GetArrayItem(rows, li);
    /* JS: line.trim(); skip empty (a 1-cell empty row) */
    if (cJSON_GetArraySize(row) <= 1) {
      const char *only = cell(row, 0);
      if (!only || !*only) continue;
    }
    const char *lonS = cell(row, iLon), *latS = cell(row, iLat);
    char *e1 = NULL, *e2 = NULL;
    double lon = lonS ? strtod(lonS, &e1) : 0;
    double lat = latS ? strtod(latS, &e2) : 0;
    int geocoded = lonS && latS && e1 != lonS && e2 != latS && *e1 == 0 && *e2 == 0;

    cJSON *f = cJSON_CreateObject();
    cJSON_AddStringToObject(f, "type", "Feature");
    if (geocoded) {
      cJSON *g = cJSON_CreateObject();
      cJSON_AddStringToObject(g, "type", "Point");
      cJSON *co = cJSON_CreateArray();
      cJSON_AddItemToArray(co, cJSON_CreateNumber(lon));
      cJSON_AddItemToArray(co, cJSON_CreateNumber(lat));
      cJSON_AddItemToObject(g, "coordinates", co);
      cJSON_AddItemToObject(f, "geometry", g);
    } else {
      cJSON_AddItemToObject(f, "geometry", cJSON_CreateNull());
    }
    cJSON *p = cJSON_CreateObject();
    cJSON_AddItemToObject(p, "bright", num_or_null(cell(row, iBright)));
    const char *acq = cell(row, iAcq);
    cJSON_AddStringToObject(p, "acq_date", acq ? acq : "");
    const char *acqt = cell(row, iAcqT);
    cJSON_AddStringToObject(p, "acq_time", acqt ? acqt : "");
    const char *conf = cell(row, iConf);
    cJSON_AddStringToObject(p, "confidence", conf ? conf : "");
    cJSON_AddItemToObject(p, "frp", num_or_null(cell(row, iFrp)));
    const char *sat = cell(row, iSat);
    cJSON_AddStringToObject(p, "satellite", sat ? sat : "");
    cJSON_AddStringToObject(p, "source", "firms_active_fire");
    cJSON_AddItemToObject(f, "properties", p);
    cJSON_AddItemToArray(features, f);
  }
  cJSON_Delete(rows);

  int n = geojson_emit_features(sink, ctx->source_id, features);
  cJSON_Delete(features);
  fprintf(stderr, "[nasa-firms-jp] emitted %d\n", n);
  return n >= 0 ? 0 : -1;
}

static const source_def nasa_firms_jp_def = {
  .id = "nasa-firms-jp", .collector = "environment",
  .name = "NASA FIRMS active fires (JP)", .name_ja = "NASA FIRMS 活火災 日本",
   .update_interval_sec = 1800, .run = run };
REGISTER_SOURCE(nasa_firms_jp_def)
